### DISCUSSION TITLE
Add to Home Screen support in iOS 16.4

### DISCUSSION THREAD

**Discussion by gutley** [mozilla-mobile/firefox-ios#13281] [General]
Not sure if explicit support is needed for this as the documentation is a bit vague as to whether the "Add to Home Screen" options is automatically added to the share sheet or not. If not automatic, it should probably be planned for as it would be nice support adding to home screen from Firefox

https://webkit.org/blog/13878/web-push-for-web-apps-on-ios-and-ipados/
and scroll to "Third-party browser support for Add to Home Screen"

"Starting in iOS and iPadOS 16.4 beta 1, Home Screen bookmarks will now open in the user’s current default browser"

**Comment by gutley** [comment]
I quickly hacked something into the code and it works correctly. The important point is: "A WKWebView is included in the array of activityItems".

So Just as a POC:

in ShareExtensionHelper: modify createActivityViewController to accept a webView as a param and to add it to the activity items:

    func createActivityViewController(webView: WKWebView? = nil, _ completionHandler: @escaping (_ completed: Bool, _ activityType: UIActivity.ActivityType?) -> Void) -> UIActivityViewController {
        var activityItems = getActivityItems(url: url)
        if webView != nil {activityItems.append(webView!)}

in BrowserViewController presentShareSheet: call createActivityViewController to pass in the tab's webview:

        let controller = helper.createActivityViewController(webView: tab?.webView, { [unowned self] completed, activityType in

That's all that's needed When you run it on ios16.4 the Add to Home Screen option is available and works.

Notes:

1. Running this in Fennec won't work due to the lack of  com.apple.developer.web-browser entitlement, but in Firefox it does.
2. In the simulator, the option to add to the home screen shows up and 'works' except nothing actually gets added to the Home Screen. I suspect this is an imitation of the simulator as the same happens in Safari. All is goos on a real device.
3. There is no great need to wait for iOS 16.4 to be released before implementing this as pre-16.4 the Webview added to the activityItems sits dormant doing nothing. You could wrap the adding to the list with a #available(iOS 16.4), but its not needed.
4. This doesn't;t need the iOS 16.4 SDK nor Xcode 14.3 beta to do this.
5. I didn't attempt to add a webview to any other calls to createActivityViewController. I just wanted to know if it would work and wasn't sure when all the other calls were made.

**Comment by jensimmons** [comment]
It's not automatically added to the Share menu. Browsers on iOS will need to add support. 

>Applications on iOS and iPadOS present the Share menu by creating a UIActivityViewController with an array of activityItems. For “Add to Home Screen” to be included in the Share menu the following must be true:
> - The application has the [com.apple.developer.web-browser](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_web-browser) managed entitlement
> - A WKWebView is included in the array of activityItems
> - The WKWebView is displaying a document with an HTTP or HTTPS URL
> - If the device is an iPad, it must not be configured as a [Shared iPad](https://support.apple.com/guide/deployment/shared-ipad-overview-dep9a34c2ba2/web)

We are happy to help if there are any questions.

**Comment by beidson** [comment]
Simulator support for A2HS is known to be buggy (In Safari, too)

But your code change seems exactly right to me to work on device (for Firefox, which has the browser entitlement)

**Reply by jensimmons** [reply]
> Simulator support for A2HS is known to be buggy (In Safari, too)

Brady means there is currently a known regression bug that causes "Add to Home Screen" to not work in any browser on iPhone and iPad Simulators running iOS 16.4 beta 1. The bug is currently being fixed. Meanwhile, testing with real devices works.
